### PR TITLE
Merge VariableTermStores and TermStores.

### PR DESCRIPTION
**Pull request by dickensc**
Merge the VariableTermStores and TermStores interfaces into a single abstract TermStore class. 
This way more code can be shared across TermStores and variable operations necessary in future weight learning applications is supported by ConsensusTermStores.

ConsensusTermStores previously had an internal reference to a MemoryVariableTermStore. This pattern is inverted so that ConsensusTermStores extend MemoryVariableTermStores with local variable data structures and methods.

MemoryVariableTermStores previously had an internal MemoryTermStore that was a wrapper for an ArrayList data structure. This functionality was merged into MemoryVariableTermStores for simplicity.

Consistent synchronization of class variables in TermStores is enforced by class methods and multi-threaded algorithms in ADMMReasoner use direct references to the TermStore data structures to support efficient parallelization / non-blocking updates.

### CODE REVIEW

**Review by cfpryor — CHANGES_REQUESTED** [review]
Should we change naming variables to atoms, do we want memoryConsensusTermStore to be absorbed into admmTermStore, simplify some file names MemoryVariableTermStore -> MemoryTermStore, and set up a memory directory inside term.

File: psl-core/src/main/java/org/linqs/psl/reasoner/admm/ADMMReasoner.java [hunk 1]
```diff
@@ -301,30 +303,30 @@ public Object clone() {
 
         @Override
         public void work(long blockIndex, Long ignore) {
-            long numTerms = termStore.size();
```
[the remark this answers is not in the document]
Comment (dickensc) [reply psl-core/src/main/java/org/linqs/psl/reasoner/admm/ADMMReasoner.java:304]: Currently no. TermWorkers are made after terms are created.

File: psl-core/src/main/java/org/linqs/psl/reasoner/admm/ADMMReasoner.java [hunk 2]
```diff
@@ -301,30 +303,30 @@ public Object clone() {
...
-
             // Minimize each local function (wrt the local variable copies).
             for (int innerBlockIndex = 0; innerBlockIndex < blockSize; innerBlockIndex++) {
-                long termIndex = blockIndex * blockSize + innerBlockIndex;
+                int termIndex = (int)(blockIndex * blockSize + innerBlockIndex);
```
[the remark this answers is not in the document]
Comment (dickensc) [reply psl-core/src/main/java/org/linqs/psl/reasoner/admm/ADMMReasoner.java:307]: Lists, which memoryTermStores, and hence ADMMTermStores, use for storing terms can only can integer indices. Before this change the term index was type casted after this operation.

File: psl-core/src/main/java/org/linqs/psl/reasoner/admm/ADMMReasoner.java [hunk 3]
```diff
@@ -335,6 +337,8 @@ public VariableWorker(ADMMTermStore termStore, long blockSize, boolean useNonCon
             this.useNonConvex = useNonConvex;
 
             this.consensusValues = termStore.getConsensusValues();
+            this.localVariables = termStore.getLocalVariables();
```
[the remark this answers is not in the document]
Comment (dickensc) [reply psl-core/src/main/java/org/linqs/psl/reasoner/admm/ADMMReasoner.java]: The reference is saved here so we call the synchronized method only once during instantiation. Then the workers can operate on the object in parallel without blocking each other by using the synchronized termStore class methods.

File: psl-core/src/main/java/org/linqs/psl/reasoner/sgd/SGDReasoner.java [hunk 4]
```diff
@@ -103,12 +102,9 @@ public SGDReasoner() {
     @Override
     public double optimize(TermStore baseTermStore,
             List<Evaluator> evaluators, TrainingMap trainingMap, Set<StandardPredicate> evaluationPredicates) {
-        if (!(baseTermStore instanceof VariableTermStore)) {
-            throw new IllegalArgumentException("SGDReasoner requires a VariableTermStore (found " + baseTermStore.getClass().getName() + ").");
-        }
 
         @SuppressWarnings("unchecked")
-        VariableTermStore<SGDObjectiveTerm, GroundAtom> termStore = (VariableTermStore<SGDObjectiveTerm, GroundAtom>)baseTermStore;
+        TermStore<SGDObjectiveTerm, GroundAtom> termStore = (TermStore<SGDObjectiveTerm, GroundAtom>)baseTermStore;
```
[the remark this answers is not in the document]
Comment (dickensc) [reply psl-core/src/main/java/org/linqs/psl/reasoner/sgd/SGDReasoner.java:107]: Not with the the current reasoner signature for optimize.

**Pull request closed without merging**